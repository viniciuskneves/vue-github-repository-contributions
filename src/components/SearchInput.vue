<template>
  <input
    class="form-control"
    type="search"
    v-model.trim="value"
    :placeholder="placeholder"
  />
</template>

<script>
let timeoutId;

export default {
  name: 'SearchInput',
  props: {
    value: {
      type: String,
      required: true,
    },
    placeholder: {
      type: String,
      required: false,
      default: '',
    },
  },
  watch: {
    value(newValue) {
      // TODO: Use request time instead of debounce as it will improve experience
      clearTimeout(timeoutId);
      timeoutId = setTimeout(() => {
        this.emitOnDebounce(newValue);
      }, 500);
    },
  },
  methods: {
    emitOnDebounce(value) {
      this.$emit('input', value);
    },
  },
};
</script>
