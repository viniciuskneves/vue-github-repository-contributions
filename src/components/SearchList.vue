<template>
  <ul class="list-group">
    <search-list-item
      v-for="repository in repositories"
      :key="repository.fullName"
      :name="repository.displayName"
      :active="repository.fullName === activeRepository.fullName"
      @click="handleItemClick(repository)"
    />
  </ul>
</template>

<script>
import { mapState } from 'vuex';
import SearchListItem from '@/components/SearchListItem.vue';

export default {
  name: 'SearchList',
  components: {
    SearchListItem,
  },
  props: {
    repositories: {
      type: Array,
      required: false,
      default: () => [],
    },
  },
  computed: mapState(['activeRepository']),
  methods: {
    handleItemClick(repository) {
      this.$store.dispatch('SEARCH_CONTRIBUTORS', repository);
    },
  },
};
</script>
