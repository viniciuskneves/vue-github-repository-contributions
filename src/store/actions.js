import api from '@/utils/api';

export default {
  async SEARCH_REPOSITORIES({ commit }, username) {
    let data = [];

    if (username) {
      data = await api.searchRepositories(username);
    }

    commit('SET_REPOSITORIES', data);

    return data;
  },
  async SEARCH_CONTRIBUTORS({ state, commit }, repository) {
    const { activeRepository } = state;
    let { contributors: data } = state;

    if (activeRepository !== repository) {
      data = await api.searchContributors(repository.fullName);

      commit('SET_ACTIVE_REPOSITORY', repository);
      commit('SET_CONTRIBUTORS', data);
    }

    return data;
  },
};
