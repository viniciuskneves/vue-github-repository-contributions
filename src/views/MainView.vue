<template>
  <div class="row">
    <div class="col-12 col-md-6 bg-light py-3 border">
      <search-input
        class="mb-2"
        placeholder="Type a GitHub username"
        @text="handleSearchUser"
      />
      <search-list
        :repositories="repositories"
      />
    </div>
    <div class="col-12 col-md-6 bg-light py-3 border">
      <repository-contributions-chart
        v-if="contributors.length && Object.keys(activeRepository).length"
        :contributors="contributors"
        :title="activeRepository.fullName"
      />
      <h3 v-else-if="!contributors.length && Object.keys(activeRepository).length">
        There are no contributions to this repository, please select another repository.
      </h3>
      <h5 v-else>
        Looks like you haven't selected a repository yet.
        Please type any GitHub username in the searchbox and then select a repository from the list.
      </h5>
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import SearchInput from '@/components/SearchInput.vue';
import SearchList from '@/components/SearchList.vue';
import RepositoryContributionsChart from '@/components/RepositoryContributionsChart.vue';

export default {
  name: 'MainView',
  components: {
    SearchInput,
    SearchList,
    RepositoryContributionsChart,
  },
  computed: mapState(['repositories', 'contributors', 'activeRepository']),
  methods: {
    handleSearchUser(username) {
      this.$store.dispatch('SEARCH_REPOSITORIES', username);
    },
  },
};
</script>
