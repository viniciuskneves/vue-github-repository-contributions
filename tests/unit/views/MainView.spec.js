import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import MainView from '@/views/MainView.vue';
import SearchInput from '@/components/SearchInput.vue';
import SearchList from '@/components/SearchList.vue';
import RepositoryContributionsChart from '@/components/RepositoryContributionsChart.vue';
import actions from '@/store/actions';
import baseState from '@/store/state';

const localVue = createLocalVue();
localVue.use(Vuex);

describe('MainView', () => {
  let wrapper;
  let state;

  beforeEach(() => {
    state = { ...baseState };
    actions.SEARCH_REPOSITORIES = jest.fn();
    wrapper = shallowMount(MainView, {
      localVue,
      store: new Vuex.Store({ state, actions }),
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders view', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders children components', () => {
    const searchInput = wrapper.find(SearchInput);
    const searchList = wrapper.find(SearchList);
    const repositoryContributionsChart = wrapper.find(RepositoryContributionsChart);
    const welcomeMessage = wrapper.find('h5');

    expect(searchInput.exists()).toBe(true);
    expect(searchList.exists()).toBe(true);
    expect(repositoryContributionsChart.exists()).toBe(false);
    expect(welcomeMessage.exists()).toBe(true);
  });

  it('renders charts only when contributors and activeRepository', () => {
    const contributors = [{
      name: 'name',
      count: 10,
    }];
    const repository = {
      displayName: 'name',
      fullName: 'fullName',
    };

    state.activeRepository = repository;
    state.contributors = contributors;

    const repositoryContributionsChart = wrapper.find(RepositoryContributionsChart);

    expect(repositoryContributionsChart.exists()).toBe(true);
  });

  it('shows message when there are no contributors', () => {
    const contributors = [];
    const repository = {
      displayName: 'name',
      fullName: 'fullName',
    };

    state.activeRepository = repository;
    state.contributors = contributors;

    const repositoryContributionsChart = wrapper.find(RepositoryContributionsChart);
    const noContributorsMessage = wrapper.find('h3');

    expect(repositoryContributionsChart.exists()).toBe(false);
    expect(noContributorsMessage.exists()).toBe(true);
  });

  it('handle searchInput emit text', () => {
    const mock = jest.fn();
    const username = 'some-github-username';
    const searchInput = wrapper.find(SearchInput);

    wrapper.setMethods({
      handleSearchUser: mock,
    });

    searchInput.vm.$emit('text', username);

    expect(mock).toHaveBeenCalledWith(username);
  });

  it('delegates "handleSearchUser" to action', () => {
    const username = 'some-github-username';
    const { vm } = wrapper;

    vm.handleSearchUser(username);

    expect(actions.SEARCH_REPOSITORIES).toHaveBeenCalled();
    expect(actions.SEARCH_REPOSITORIES.mock.calls[0][1]).toBe(username);
  });

  it('binds "state.repositories" to searchList', () => {
    const repositories = [{
      displayName: 'angular',
      fullName: 'angular/angular',
    }];

    state.repositories = repositories;

    const searchList = wrapper.find(SearchList);

    expect(searchList.vm.repositories).toBe(repositories);
  });

  it('binds "state.contributors" to chart', () => {
    const repository = {
      displayName: 'name',
      fullName: 'fullName',
    };
    const contributors = [{
      name: 'name',
      count: 10,
    }];

    state.contributors = contributors;
    state.activeRepository = repository;

    const repositoryContributionsChart = wrapper.find(RepositoryContributionsChart);

    expect(repositoryContributionsChart.vm.contributors).toBe(contributors);
  });

  it('binds "state.activeRepository.fullName" to chart title', () => {
    const repository = {
      displayName: 'name',
      fullName: 'fullName',
    };
    const contributors = [{
      name: 'name',
      count: 10,
    }];

    state.contributors = contributors;
    state.activeRepository = repository;

    const repositoryContributionsChart = wrapper.find(RepositoryContributionsChart);

    expect(repositoryContributionsChart.vm.title).toBe(repository.fullName);
  });
});
