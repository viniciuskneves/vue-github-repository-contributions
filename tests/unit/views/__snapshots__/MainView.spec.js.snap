// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainView renders view 1`] = `
<div class="row">
  <div class="col-12 col-md-6 bg-light py-3 border">
    <searchinput-stub placeholder="Type a GitHub username" class="mb-2"></searchinput-stub>
    <searchlist-stub repositories=""></searchlist-stub>
  </div>
  <div class="col-12 col-md-6 bg-light py-3 border">
    <h5>
      Looks like you haven't selected a repository yet.
      Please type any GitHub username in the searchbox and then select a repository from the list.
    </h5>
  </div>
</div>
`;
